File: rb/test/weather_entity_test.rb
# Weather entity test

require "minitest/autorun"
require "json"
require_relative "../Weather_sdk"
require_relative "runner"

class WeatherEntityTest < Minitest::Test
  def test_create_instance
    testsdk = WeatherSDK.test(nil, nil)
    ent = testsdk.Weather(nil)
    assert !ent.nil?
  end

  def test_basic_flow
    setup = weather_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["load"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "weather." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set WEATHER_TEST_WEATHER_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # Bootstrap entity data from existing test data.
    weather_ref01_data_raw = Vs.items(Helpers.to_map(
      Vs.getpath(setup[:data], "existing.weather")))
    weather_ref01_data = nil
    if weather_ref01_data_raw.length > 0
      weather_ref01_data = Helpers.to_map(weather_ref01_data_raw[0][1])
    end

    # LOAD
    weather_ref01_ent = client.Weather(nil)
    weather_ref01_match_dt0 = {}
    weather_ref01_data_dt0_loaded, err = weather_ref01_ent.load(weather_ref01_match_dt0, nil)
    assert_nil err
    assert !weather_ref01_data_dt0_loaded.nil?

  end
end

def weather_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "weather", "WeatherTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = WeatherSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["weather01", "weather02", "weather03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["WEATHER_TEST_WEATHER_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "WEATHER_TEST_WEATHER_ENTID" => idmap,
    "WEATHER_TEST_LIVE" => "FALSE",
    "WEATHER_TEST_EXPLAIN" => "FALSE",
    "WEATHER_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["WEATHER_TEST_WEATHER_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["WEATHER_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["WEATHER_APIKEY"],
      },
      extra || {},
    ])
    client = WeatherSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["WEATHER_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["WEATHER_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end

File: go/test/weather_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/weather-sdk/go"
	"github.com/voxgig-sdk/weather-sdk/go/core"

	vs "github.com/voxgig-sdk/weather-sdk/go/utility/struct"
)

func TestWeatherEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Weather(nil)
		if ent == nil {
			t.Fatal("expected non-nil WeatherEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := weatherBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "weather." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set WEATHER_TEST_WEATHER_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		weatherRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.weather", setup.data)))
		var weatherRef01Data map[string]any
		if len(weatherRef01DataRaw) > 0 {
			weatherRef01Data = core.ToMapAny(weatherRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = weatherRef01Data

		// LOAD
		weatherRef01Ent := client.Weather(nil)
		weatherRef01MatchDt0 := map[string]any{
			"id": weatherRef01Data["id"],
		}
		weatherRef01DataDt0Loaded, err := weatherRef01Ent.Load(weatherRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		weatherRef01DataDt0LoadResult := core.ToMapAny(entityData(weatherRef01DataDt0Loaded))
		if weatherRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if weatherRef01DataDt0LoadResult["id"] != weatherRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func weatherBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "weather", "WeatherTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read weather test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse weather test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"weather01", "weather02", "weather03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("WEATHER_TEST_WEATHER_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"WEATHER_TEST_WEATHER_ENTID": idmap,
		"WEATHER_TEST_LIVE":      "FALSE",
		"WEATHER_TEST_EXPLAIN":   "FALSE",
	})

	idmapResolved := core.ToMapAny(env["WEATHER_TEST_WEATHER_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["WEATHER_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
			},
			extra,
		})
		client = sdk.NewWeatherSDK(core.ToMapAny(mergedOpts))
	}

	live := env["WEATHER_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["WEATHER_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
